broker: relocate temporary statedir to /var/tmp unless overridden

garlick opened the pull request:
As noted in #6996, `/var/tmp` is a better default location for the KVS content db than `/tmp`.

When `rundir` and `statedir` are not overridden on the command line, `rundir` is a directory in `/tmp` by default, and `statedir` is unset, causing `content-sqlite` to use `rundir` for its database.  This PR changes the broker to create a separate `statedir`in `/var/tmp` by default.

Note that if TMPDIR is set, both directories are created there, following [systemd guidance](https://systemd.io/TEMPORARY_DIRECTORIES/).

This does not affect the system instance since both directories are set on the command line by the systemd unit file.

To make it possible for `content-sqlite` and the dump logic in the rc/modprobe scripts to know whether `statedir` is a temporary directory or a persistent one, a new `statedir-cleanup` attribute is added which is set to 1 if the directory will be removed on exit.  (There is also `rundir-cleanup` now FWIW).

The code is structured to accomodate a follow-on PR to add an option to suggest other tmp-like locations where a temporary statedir directory might be created, such as `/l/ssd` on the coral2 systems as discussed in #6990

grondo commented on doc/man7/flux-broker-attributes.rst:
```diff
@@ -34,22 +34,34 @@ version
+   For example, content backing store data is stored here to facilitate
+   restarts.  If unset, a unique temporary directory is created that
+   is cleaned up on instance shutdown.
+   If set, this directory must exist and be owned by the instance owner.
+   The state directory is only used on the leader (rank 0) broker.
```
Is this correct? It seems like the broker will create the last component of `statedir`
```console
$ flux batch -N1 --output=test.out --broker-opts=-Sstatedir=/var/tmp/grondo/state-{{id}} --wrap flux getattr statedir
f23zjhCzB
$ cat test.out
/var/tmp/grondo/state-f23zjhCzB
$ flux batch -N1 --output=test.out --broker-opts=-Sstatedir=/var/tmp/grondo/flux/state-{{id}} --wrap flux getattr statedir
f4RHpTdif
$ cat test.out
lt-flux-broker: statedir error creating /var/tmp/grondo/flux/state-f4RHpTdif: No such file or directory
```

garlick replied:
Yep the docs are not correct - good catch!

The broker also removes the directory if it creates it.  Now I'm thinking that's wrong, it should only remove it if the directory was not specified and it came from `mkdtemp()`.

grondo replied:
> The broker also removes the directory if it creates it. Now I'm thinking that's wrong, it should only remove it if the directory was not specified and it came from mkdtemp().

Ok, and we'd rely on other cleanup mechanisms in the case of e.g. `-Sstatedir=/l/ssd/{{jobid}}`?

garlick replied:
I actually wasn't thinking about this for that use case because {{jobid}} cannot be guaranteed to be unique.  I was thinking instead to add a follow-up PR that allowed another tmpdir like directory (like `/l/ssd`) to be specified in another option and then let the broker would use` mkdtemp()` in that instead of in TMPDIR or /var/tmp.

My concern about removing a directory that the user actually named is that they might expect it to persist with its contents.

grondo replied:
That makes sense. Note that in the use case specified in the original issue {{jobid}} would be guaranteed to be unique because the user knows they aren't running more subinstances within their batch job (or those do not need a manually specified `statedir`)

However, having a way to easily ask for a `statedir` to be created in a different filesystem than the default is a much more user friendly solution!

grondo reviewed (approved):
This LGTM. Just had one question about the docs.

I did test that template substitution works for broker opts (as seen in the comment below).

garlick commented:
OK, implemented that and expanded testing.  I also fixed a problem where two of the local-uri tests were failing in the fedora40 clang-18 CI builder (only) due to a very long trash directory path.

garlick commented:
I'm seeing the t2492-shell-lost.sh test fail quite a bit.  It's one that was recently tweaked in 6e173bf84976f00c21fd36addae8c9883897dce4 so I thought i'd mention it here.  I'll restart that builder for nw.
```
2025-08-29T19:23:11.1017808Z 
2025-08-29T19:23:11.1017902Z expecting success: flux job attach fQbPa7J
2025-08-29T19:23:11.1018171Z flux-job: task(s) exited with exit code 1
2025-08-29T19:23:11.1018532Z t2492: Re-launching test script under flux-start using /usr/src/src/cmd/flux
2025-08-29T19:23:11.1018864Z t2492: 
2025-08-29T19:23:11.1019054Z t2492: Testing handling of lost shell rank 3:
2025-08-29T19:23:11.1019380Z t2492: Submitted job fWz2UZV. Waiting for shell rank 3 to be lost
2025-08-29T19:23:11.1020209Z [1m[32m[Aug29 19:22][0m [33mheader[0m [34mversion[0m=[37m1[0m [34mencoding[0m=[35m{"stdout":"UTF-8","stderr":"UTF-8"}[0m [34mcount[0m=[35m{"stdout":4,"stderr":4}[0m [34moptions[0m=[35m{}[0m
2025-08-29T19:23:11.1021413Z [32m[  +0.450690][0m [33mlog[0m [34mrank[0m=[37m0[0m [34mlevel[0m=[37m4[0m [34mmessage[0m=[35m"shell rank 3 (on f2cb9d5f91ce): Killed"[0m [34mcomponent[0m=[35m"exception"[0m [34mfile[0m=[35m"exception.c"[0m [34mline[0m=[37m49[0m
2025-08-29T19:23:11.1022148Z t2492: Sending SIGINT to fWz2UZV. Job should now exit
2025-08-29T19:23:11.1022503Z 0.000s: job.submit {"userid":1001,"urgency":16,"flags":0,"version":1}
2025-08-29T19:23:11.1022812Z 0.014s: job.validate
2025-08-29T19:23:11.1022996Z 0.026s: job.depend
2025-08-29T19:23:11.1023446Z 0.026s: job.priority {"priority":16}
2025-08-29T19:23:11.1023799Z 0.032s: job.alloc {"annotations":{"sched":{"resource_summary":"rank[0-3]/core0"}}}
2025-08-29T19:23:11.1024147Z 0.053s: job.start
2025-08-29T19:23:11.1024329Z flux-job: job shell Killed
2025-08-29T19:23:11.1024541Z 0.691s: job.finish {"status":9}
2025-08-29T19:23:11.1024781Z 0.698s: job.release {"ranks":"all","final":true}
2025-08-29T19:23:11.1025038Z 0.698s: job.free
2025-08-29T19:23:11.1025201Z 0.698s: job.clean
2025-08-29T19:23:11.1025369Z 0.033s: exec.init
2025-08-29T19:23:11.1025534Z 0.038s: exec.starting
2025-08-29T19:23:11.1025836Z 0.142s: exec.shell.init {"service":"1001-shell-fWz2UZV","leader-rank":0,"size":4}
2025-08-29T19:23:11.1026247Z 0.150s: exec.shell.start {"taskmap":{"version":1,"map":[[0,4,1,1]]}}
2025-08-29T19:23:11.1026770Z 0.662s: exec.shell.task-exit {"localid":0,"rank":0,"state":"Exited","pid":246713,"wait_status":0,"signaled":0,"exitcode":0}
2025-08-29T19:23:11.1027255Z 0.691s: exec.complete {"status":9}
2025-08-29T19:23:11.1027477Z 0.691s: exec.done
2025-08-29T19:23:11.1027809Z 0.570s: flux-shell[0]:  WARN: exception: exception.c:49: shell rank 3 (on f2cb9d5f91ce): Killed
2025-08-29T19:23:11.1028325Z waiting on barrier for 4 tasks
2025-08-29T19:23:11.1028560Z flux-pmi: operation failed
2025-08-29T19:23:11.1028759Z exited barrier
2025-08-29T19:23:11.1028943Z flux-pmi: operation failed
2025-08-29T19:23:11.1029146Z flux-pmi: operation failed
2025-08-29T19:23:11.1029392Z t2492: Job exited with rc=137 (expecting 137 (128+9))
2025-08-29T19:23:11.1029656Z t2492: 
2025-08-29T19:23:11.1029840Z t2492: Testing handling of lost shell rank 1:
2025-08-29T19:23:11.1030171Z t2492: Submitted job fwYkaNs. Waiting for shell rank 1 to be lost
2025-08-29T19:23:11.1030983Z [1m[32m[Aug29 19:22][0m [33mheader[0m [34mversion[0m=[37m1[0m [34mencoding[0m=[35m{"stdout":"UTF-8","stderr":"UTF-8"}[0m [34mcount[0m=[35m{"stdout":4,"stderr":4}[0m [34moptions[0m=[35m{}[0m
2025-08-29T19:23:11.1032300Z [32m[  +0.293673][0m [33mlog[0m [34mrank[0m=[37m0[0m [34mlevel[0m=[37m4[0m [34mmessage[0m=[35m"shell rank 1 (on f2cb9d5f91ce): Killed"[0m [34mcomponent[0m=[35m"exception"[0m [34mfile[0m=[35m"exception.c"[0m [34mline[0m=[37m49[0m
2025-08-29T19:23:11.1033214Z t2492: Sending SIGINT to fwYkaNs. Job should now exit
2025-08-29T19:23:11.1033574Z 0.000s: job.submit {"userid":1001,"urgency":16,"flags":0,"version":1}
2025-08-29T19:23:11.1033880Z 0.014s: job.validate
2025-08-29T19:23:11.1034060Z 0.027s: job.depend
2025-08-29T19:23:11.1034254Z 0.027s: job.priority {"priority":16}
2025-08-29T19:23:11.1034590Z 0.030s: job.alloc {"annotations":{"sched":{"resource_summary":"rank[0-3]/core0"}}}
2025-08-29T19:23:11.1034930Z 0.047s: job.start
2025-08-29T19:23:11.1035101Z 0.031s: exec.init
2025-08-29T19:23:11.1035269Z 0.033s: exec.starting
2025-08-29T19:23:11.1035570Z 0.134s: exec.shell.init {"service":"1001-shell-fwYkaNs","leader-rank":0,"size":4}
2025-08-29T19:23:11.1035982Z 0.146s: exec.shell.start {"taskmap":{"version":1,"map":[[0,4,1,1]]}}
2025-08-29T19:23:11.1036500Z 0.559s: exec.shell.task-exit {"localid":0,"rank":3,"state":"Exited","pid":247538,"wait_status":2,"signaled":2,"exitcode":130}
2025-08-29T19:23:11.1036962Z 0.601s: exec.complete {"status":33280}
2025-08-29T19:23:11.1037192Z 0.601s: exec.done
2025-08-29T19:23:11.1037513Z 0.422s: flux-shell[0]:  WARN: exception: exception.c:49: shell rank 1 (on f2cb9d5f91ce): Killed
2025-08-29T19:23:11.1037899Z Traceback (most recent call last):
2025-08-29T19:23:11.1038250Z   File "/tmp/flux-BgnJlD/jobtmp-3-fwYkaNs/critical.py", line 4, in <module>
2025-08-29T19:23:11.1038587Z     import flux
2025-08-29T19:23:11.1038883Z   File "/usr/src/src/bindings/python/flux/__init__.py", line 14, in <module>
2025-08-29T19:23:11.1039232Z     import flux.core.handle
2025-08-29T19:23:11.1039560Z   File "/usr/src/src/bindings/python/flux/core/handle.py", line 19, in <module>
2025-08-29T19:23:11.1039925Z     from flux.future import Future
2025-08-29T19:23:11.1040247Z   File "/usr/src/src/bindings/python/flux/future.py", line 16, in <module>
2025-08-29T19:23:11.1040670Z     from flux.util import check_future_error, interruptible
2025-08-29T19:23:11.1041049Z   File "/usr/src/src/bindings/python/flux/util.py", line 44, in <module>
2025-08-29T19:23:11.1041411Z     from flux.utils.parsedatetime import Calendar
2025-08-29T19:23:11.1041854Z   File "/usr/src/src/bindings/python/flux/utils/parsedatetime/__init__.py", line 40, in <module>
2025-08-29T19:23:11.1042300Z     from .warns import pdt20DeprecationWarning
2025-08-29T19:23:11.1042623Z   File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
2025-08-29T19:23:11.1043180Z   File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
2025-08-29T19:23:11.1043634Z   File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
2025-08-29T19:23:11.1044026Z   File "<frozen importlib._bootstrap_external>", line 936, in exec_module
2025-08-29T19:23:11.1044439Z   File "<frozen importlib._bootstrap_external>", line 1069, in get_code
2025-08-29T19:23:11.1044853Z   File "<frozen importlib._bootstrap_external>", line 729, in _compile_bytecode
2025-08-29T19:23:11.1045192Z KeyboardInterrupt
2025-08-29T19:23:11.1045501Z flux-job: task(s) Interrupt
2025-08-29T19:23:11.1045721Z 0.601s: job.finish {"status":33280}
2025-08-29T19:23:11.1045974Z 0.608s: job.release {"ranks":"all","final":true}
2025-08-29T19:23:11.1046217Z 0.608s: job.free
2025-08-29T19:23:11.1046380Z 0.608s: job.clean
2025-08-29T19:23:11.1046587Z t2492: Job exited with rc=130 (expecting 137 (128+9))
2025-08-29T19:23:11.1046860Z t2492: Unexpected job exit code 130
2025-08-29T19:23:11.1047243Z Aug 29 19:22:42.626936 UTC 2025 broker.err[0]: rc2.0: /usr/src/t/issues/t2492-shell-lost.sh Exited (rc=1) 2.0s
2025-08-29T19:23:11.1047656Z flux-start: 0 (pid 245511) exited with rc=1
2025-08-29T19:23:11.1048048Z Aug 29 19:22:44.285609 UTC 2025 broker.err[0]: rc2.0: /usr/src/t/issues/t2492-shell-lost.sh Exited (rc=1) 4.8s
2025-08-29T19:23:11.1048591Z not ok 20 - t2492-shell-lost
```

grondo commented:
> I'm seeing the t2492-shell-lost.sh test fail quite a bit. It's one that was recently tweaked in https://github.com/flux-framework/flux-core/commit/6e173bf84976f00c21fd36addae8c9883897dce4 so I thought i'd mention it here.

Hm, maybe that wasn't the right fix because the job (`critical.py`) is getting still getting sent SIGINT early. I'll work on improving that.

grondo commented:
Ah ha:
```console
$ flux run -N4 flux python -c 'import subprocess; subprocess.run(["flux", "pmi", "barrier"]);'
ƒ8kGBt4Qf: completed pmi barrier on 1 tasks in 0.000s.
ƒ8kGBt4Qf: completed pmi barrier on 1 tasks in 0.000s.
ƒ8kGBt4Qf: completed pmi barrier on 1 tasks in 0.000s.
ƒ8kGBt4Qf: completed pmi barrier on 1 tasks in 0.000s.
$ flux run -N4 flux python -c 'import subprocess; subprocess.run(["flux", "pmi", "barrier"], close_fds=False);'
ƒ8oeRNzqu: completed pmi barrier on 4 tasks in 0.001s.
```
D'oh!

grondo reviewed (approved):
Still LGTM!
